File: Objective C/Auropay ObjC Demo/Auropay ObjC Demo/ViewController.m
//
//  ViewController.m
//  Auropay ObjC Demo
//
//  Created by AuroPay on 05/07/23.
//

#import "ViewController.h"
#import <AuroPayPayments/AuroPayPayments-Swift.h>

@interface ViewController ()

@property (weak, nonatomic) IBOutlet UITextField *textFieldAmount;

@property (weak, nonatomic) IBOutlet UITextField *textFieldFirstName;

@property (weak, nonatomic) IBOutlet UITextField *textFieldLastName;

@property (weak, nonatomic) IBOutlet UITextField *textFieldEmail;

@property (weak, nonatomic) IBOutlet UITextField *textFieldPhone;


@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.

    _textFieldAmount.leftView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 15, _textFieldAmount.frame.size.height)];
    _textFieldAmount.leftViewMode = UITextFieldViewModeAlways;
    
    _textFieldFirstName.leftView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 15, _textFieldFirstName.frame.size.height)];
    _textFieldFirstName.leftViewMode = UITextFieldViewModeAlways;
    
    _textFieldLastName.leftView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 15, _textFieldLastName.frame.size.height)];
    _textFieldLastName.leftViewMode = UITextFieldViewModeAlways;
    
    _textFieldEmail.leftView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 15, _textFieldEmail.frame.size.height)];
    _textFieldEmail.leftViewMode = UITextFieldViewModeAlways;
    
    _textFieldPhone.leftView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 15, _textFieldPhone.frame.size.height)];
    _textFieldPhone.leftViewMode = UITextFieldViewModeAlways;
    
}


- (IBAction)onTapPay:(UIButton *)sender {
    
    //Create Theme
    APTheme *theme = [[APTheme alloc] initWithPrimaryColor:[self colorFromHexString:@"#236C6C"]
                     primaryTextColor:[UIColor whiteColor]
                     secondaryColor:[UIColor redColor]
                     secondaryTextColor:[UIColor whiteColor]];
    
    //Initilize the AuropayBuilder with subdomain Id, accessKey and secretKey
    Auropay * apb = [[[[[[[[[AuroPayBuilder alloc]
                            merchantID:@"Your Subdomain ID"]
                            accessKey:@"Your Access Key"]
                            secretKey:@"Your Secret Key"]
                            customerProfile:[self getCustomerProfile]]
                            theme:theme]
                            showReceipt:YES]
                            allowCardScan:YES]
                            build];

    //Initiate payment
    if (_textFieldAmount.text && _textFieldAmount.text.length > 0){
        [apb doPaymentWithDisplayViewController:self amount:[_textFieldAmount.text doubleValue] andDelegate:self orderId:@""];
    }
}


//Create customer profile
-(CustomerProfile *)getCustomerProfile{
    CustomerProfile *cp = [[CustomerProfile alloc] initWithTitle:@"Some Title" firstName:self.textFieldFirstName.text middleName:@"" lastName:self.textFieldLastName.text phone:self.textFieldPhone.text email:self.textFieldEmail.text];
    return  cp;
}


//Delegate methods to receive callbacks and events
-(void)onPaymentSuccess:(NSString *)orderId transactionStatus:(NSInteger)transactionStatus transactionId:(NSString *)transactionId{
    NSLog(@"orderId:- %@",orderId);
    NSLog(@"transactionStatus:- %ld",(long)transactionStatus);
    NSLog(@"transactionId:- %@",transactionId);
}

- (void)onPaymentError:(NSString *)message{
    NSLog(@"%@",message);
}


///UIColor hex to color
-(UIColor *)colorFromHexString:(NSString *)hexString {
    unsigned rgbValue = 0;
    NSScanner *scanner = [NSScanner scannerWithString:hexString];
    [scanner setScanLocation:1]; // bypass '#' character
    [scanner scanHexInt:&rgbValue];
    return [UIColor colorWithRed:((rgbValue & 0xFF0000) >> 16)/255.0 green:((rgbValue & 0xFF00) >> 8)/255.0 blue:(rgbValue & 0xFF)/255.0 alpha:1.0];
}

@end

File: Swift/Auropay Swift DemoApp/Auropay Swift DemoApp/ViewController.swift
//
//  ViewController.swift
//  Auropay Swift DemoApp
//
//  Created by AuroPay on 04/07/23.
//

import UIKit
import AuroPayPayments

class ViewController: UIViewController {

    //MARK: - IBOutlets
    @IBOutlet weak var textFieldAmount: UITextField!{
        didSet{
            guard let textField = textFieldAmount else { return }
            textField.leftView = UIView(frame: CGRect(x: 0, y: 0, width: 15, height: textField.frame.height))
            textField.leftViewMode = .always
        }
    }
    
    @IBOutlet weak var textFieldFirstName: UITextField!{
        didSet{
            guard let textField = textFieldFirstName else { return }
            textField.leftView = UIView(frame: CGRect(x: 0, y: 0, width: 15, height: textField.frame.height))
            textField.leftViewMode = .always
        }
    }
    
    @IBOutlet weak var textFieldLastName: UITextField!{
        didSet{
            guard let textField = textFieldLastName else { return }
            textField.leftView = UIView(frame: CGRect(x: 0, y: 0, width: 15, height: textField.frame.height))
            textField.leftViewMode = .always
        }
    }
    
    @IBOutlet weak var textFieldEmail: UITextField!{
        didSet{
            guard let textField = textFieldEmail else { return }
            textField.leftView = UIView(frame: CGRect(x: 0, y: 0, width: 15, height: textField.frame.height))
            textField.leftViewMode = .always
        }
    }
    
    @IBOutlet weak var textFieldPhone: UITextField!{
        didSet{
            guard let textField = textFieldPhone else { return }
            textField.leftView = UIView(frame: CGRect(x: 0, y: 0, width: 15, height: textField.frame.height))
            textField.leftViewMode = .always
        }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
    }

    @IBAction func onTapPay(_ sender: UIButton) {
        
        //Create Theme
        let theme = APTheme(primaryColor: UIColor.init(hexString: "#236C6C"), primaryTextColor: .white, secondaryColor: .red, secondaryTextColor: .white)
        
        //Initilize the AuropayBuilder with subdomain Id, accessKey and secretKey
        let apb = AuroPayBuilder()
            .merchantID("Your Subdomain ID")
            .accessKey("Your Access Key")
            .secretKey("Your Secret Key")
            .customerProfile(self.getCustomerProfile())
            .theme(theme)
            .showReceipt(true)
            .allowCardScan(true)
            .build()
        
        //Initiate payment
        if let amount = self.textFieldAmount.text, !amount.isEmpty{
            apb.doPayment(displayViewController: self, amount: (amount as NSString).doubleValue, andDelegate: self)
        }
        
    }
 
    
    //Create customer profile
    private func getCustomerProfile() -> CustomerProfile{
        let cp = CustomerProfile(title: "Some Title", firstName: self.textFieldFirstName.text ?? "", middleName: "", lastName: self.textFieldLastName.text ?? "", phone: self.textFieldPhone.text ?? "", email: self.textFieldEmail.text ?? "")
        return cp
    }
}

//Delegate methods to receive callbacks and events
extension ViewController : APPaymentCompletionProtocol{
    func onPaymentSuccess(_ orderId: String, transactionStatus: Int, transactionId: String) {
        print("orderId:- \(orderId)")
        print("transactionStatus:- \(transactionStatus)")
        print("transactionId:- \(transactionId)")
    }
    
    func onPaymentError(_ message: String) {
        print("Error:- \(message)")
    }
}




extension UIColor {
    convenience init(hexString: String) {
            let hex = hexString.trimmingCharacters(in: CharacterSet.alphanumerics.inverted)
            var int = UInt64()
            Scanner(string: hex).scanHexInt64(&int)
            let a, r, g, b: UInt64
            switch hex.count {
            case 3: // RGB (12-bit)
                (a, r, g, b) = (255, (int >> 8) * 17, (int >> 4 & 0xF) * 17, (int & 0xF) * 17)
            case 6: // RGB (24-bit)
                (a, r, g, b) = (255, int >> 16, int >> 8 & 0xFF, int & 0xFF)
            case 8: // ARGB (32-bit)
                (a, r, g, b) = (int >> 24, int >> 16 & 0xFF, int >> 8 & 0xFF, int & 0xFF)
            default:
                (a, r, g, b) = (255, 0, 0, 0)
            }
            self.init(red: CGFloat(r) / 255, green: CGFloat(g) / 255, blue: CGFloat(b) / 255, alpha: CGFloat(a) / 255)
        }
}
